### ISSUE TITLE
Complete Performance PDF page

### ISSUE THREAD

**Issue by ehharvey** [ehharvey/CSCN73060-Group6#9]
[Baseline Performance Measurements.pdf](https://github.com/ehharvey/CSCN73060-Group6/files/10511972/Baseline.Performance.Measurements.pdf)

# Steps
- [x] Define what we're measuring for each workload. For example, workload 1 (reading the file) could be defined as the average latency (or time it takes) for the client to read a line over the course of program
- [x] Implement a way to measure the latency described above. For example, we could time the start and end of when the client reads a line and store that somewhere. At the end of the program, we can average the times.
- [x] Integrate implementation with #14

## Assignment
Jacky: Workload 2
Amanuel: Workload 3
Dominic: Workload 4
Emil: Workload 5

Everyone: Workload 1

**Comment by ehharvey** [comment]
Early thoughts:
- We can possibly use `coz` profiler #13
- We can focus on latency for now. This might make things simpler

**Comment by ehharvey** [comment]
We know about the coz profiler, but others exist. For example: https://github.com/wolfpld/tracy

**Comment by ehharvey** [comment]
How to store the data:

We have a folder called performance_data (or something).

`performance_data/[profiler]/[date + time]/[workload]/*.json`

Each json (or whatever format) would contain measurement (latency or throughput, etc.)
```
{
    "latency": "3ms"
}
```

**Comment by ehharvey** [comment]
Workload 5: The server calculate the average of a category of values for every new value it receives

We will define this as the latency starting from just before the server access memory relating to stored numerical data (just before line 180). Latency measurement will end after the average is derived and stored in a variable (just after line 183).

Like other workloads, we will measure this latency using a simple profiler class we developed. A single `LatencyTimer` object will persist throughout the execution of the program. We will use a `LatencyMeasurement` object and initialize it just before line 180. We will call its `stop()` method just after line 183. We will then add this object to the `LatencyTimer` singleton for in-memory storage. At the end of the program, the `LatencyTimer` object will write its stored measurements to disk for us to review.

The `LatencyTimer` and `LatencyMeasurement`-related code will operate behind compiler directives, specifically when `TimeLatency` is `#define`d. This means that latency-timing code will not be included in the application when the macro is not defined, allowing us to measure other metrics (in other executions) without latency-related code affecting those measurements.

**Comment by Dom17Pham** [comment]
Workload 4: The server updates its data set in the memory with newly received values.

The workload will be measured for its latency which will be defined as the time it takes for the update function to allocate a new array in memory and to copy the values of the previous array into the new array including the new appended value. The execution time will be measured in milliseconds and will be measured starting from the time the update function is called by the server until the values are stored in the new array. 

The latency of this workload will be measured using a simple profiler class we developed which will allow us to create ```LatencyTimer``` and ```LatencyMeasurement``` objects to measure the latency of the workload. The measurement will start when the update function has been called by the server, starting the timer inside the scope of the update function and stopping the timer before the app leaves the scope of the update function. This will account for each time the server calls the update function as it is called multiple times for each parameter of the data set.

The results of the measurements will be stored into disk in an organized folder system that will be organized by the profiler used, date & time of when the measurements were conducted, and workload that was measured.

**Comment by ehharvey** [comment]
Workload 1: The client reads flight data from file

This workload will be measured by its latency that the Client application takes in order to read a line from the file. Typically, this metric will be viewed as the average latency (that is, the average of all latencies during a single execution).

The latency of this workload will be measured using a simple profiler class we developed which will allow us to create LatencyTimer and LatencyMeasurement objects to measure the latency of the workload. The measurement will begin when the Client application begins to seek a line, which occurs just before line 51. The measurement concludes when the Client application has retrieved a line and has it stored in memory, which occurs just after line 54.

**Comment by Amanuel-Negussie** [comment]
Workload 3: Server Determining the Category of the Flight

We can measure the latency of the workload (time it takes to determine the category of the flight --> timed from after 'ACK' has been sent to client, to after the category of flight is determined (application enters into correct if statement body. Example attached in image for category "ACCELERATION BODY Y". 


The latency of the workload will be measured by our own custom profiler which will time latency of workloads. As described above this will be implemented by the `LatencyTimer` and `LatencyMeasurement` objects developed by us in branch `rec-performance`


Image: Example of where we would call start() and end() in the respective `LatencyTimer` object instantiated for finding category "ACCELERATION BODY Y".

![screenshot](https://user-images.githubusercontent.com/74369791/215348766-dddde067-4e9d-4a8d-9b7d-75652b14be32.png)

**Comment by Jacky-Lam-CY** [comment]
Workload 2: The data communication between the client and server program.
Data communication happens on both client and server; however, the communication is initiated by the client and also ends with the client receiving the result packet. We defined one network transaction to be { Client sends heading -> Server reply ACK -> Client sends value -> Server reply with the average }. There is a special case where the server will receive a “*” sent by client indicating end of communication, but since that only happen once per session, it can be ignored.
The metric we decided to measure for data communication is the round tip time (RTT). To get the RTT of the data communication, we will measure the time for one transaction in the client program by using the LatencyMeasurement object. The LatencyMeasurement object initializes just before the first send() function and calls the stop() function of the object right after the recv() function .
